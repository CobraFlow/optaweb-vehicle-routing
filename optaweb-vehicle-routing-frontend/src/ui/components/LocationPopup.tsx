/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { Popup } from 'react-leaflet';
import { ActionGroup, Button, Form, FormGroup, TextInput } from '@patternfly/react-core';
import { Location } from '../../store/route/types';

export interface Props {
  location: Location;
  removeHandler: (id: number) => void;
  updateHandler: (location: Location) => void;
  cancelHandler: () => void;
}

export interface State {
  description?: string;
}

class LocationPopup extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      description: props.location.description,
    };

    this.handleSave = this.handleSave.bind(this);
    this.handleRemove = this.handleRemove.bind(this);
    this.handleCancel = this.handleCancel.bind(this);
  }

  handleSave() {
    const location: Location = {
      id: this.props.location.id,
      lat: this.props.location.lat,
      lng: this.props.location.lng,
      description: this.state.description,
    };
    this.props.updateHandler(location);
  }

  handleRemove() {
    this.props.removeHandler(this.props.location.id);
  }

  handleCancel() {
    this.props.cancelHandler();
  }

  render() {
    const { description } = this.state;
    const { id } = this.props.location;

    return (
      <Popup
        minWidth={300}
      >
        <span>
          {`id ${id}, description=${description}`}
        </span>
        <br />
        <Form>
          <FormGroup
            fieldId="popup-form"
            label="Description"
          >
            <TextInput
              id="input-description"
              className="popup-input"
              type="text"
              value={description}
              onChange={(value) => {
                this.setState({ description: value });
              }}
            />
          </FormGroup>
          <ActionGroup>
            <Button id="save-button" variant="primary" onClick={this.handleSave}>Save</Button>
            <Button id="remove-button" variant="link" onClick={this.handleRemove}>Remove</Button>
            <Button id="cancel-button" variant="link" onClick={this.handleCancel}>Cancel</Button>
          </ActionGroup>
        </Form>
      </Popup>
    );
  }
}

export default LocationPopup;
