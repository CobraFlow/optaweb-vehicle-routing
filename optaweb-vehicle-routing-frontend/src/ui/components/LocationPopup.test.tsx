/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { Location } from 'store/route/types';
import LocationPopup, { Props } from './LocationPopup';

const location: Location = {
  id: 1,
  lat: 1.345678,
  lng: 1.345678,
  description: 'location1',
};

describe('Location Popup', () => {
  it('render description', () => {
    const props: Props = {
      removeHandler: jest.fn(),
      updateHandler: jest.fn(),
      cancelHandler: jest.fn(),
      location,
    };
    const locationPopup = shallow(<LocationPopup {...props} />);
    expect(toJson(locationPopup)).toMatchSnapshot();
  });

  it('should call update handler when clicked', () => {
    const props: Props = {
      removeHandler: jest.fn(),
      updateHandler: jest.fn(),
      cancelHandler: jest.fn(),
      location,
    };
    const locationPopup = shallow(<LocationPopup {...props} />);
    locationPopup.find('#save-button').simulate('click');
    expect(props.updateHandler).toBeCalled();
  });

  it('should call remove handler when clicked', () => {
    const props: Props = {
      removeHandler: jest.fn(),
      updateHandler: jest.fn(),
      cancelHandler: jest.fn(),
      location,
    };
    const locationPopup = shallow(<LocationPopup {...props} />);
    locationPopup.find('#remove-button').simulate('click');
    expect(props.removeHandler).toBeCalled();
  });

  it('should call cancel handler when clicked', () => {
    const props: Props = {
      removeHandler: jest.fn(),
      updateHandler: jest.fn(),
      cancelHandler: jest.fn(),
      location,
    };
    const locationPopup = shallow(<LocationPopup {...props} />);
    locationPopup.find('#cancel-button').simulate('click');
    expect(props.cancelHandler).toBeCalled();
  });
});
