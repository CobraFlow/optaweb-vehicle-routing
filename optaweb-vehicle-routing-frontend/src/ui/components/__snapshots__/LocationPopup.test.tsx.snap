// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Location Popup render description 1`] = `
<ForwardRef(Leaflet(Popup))
  minWidth={300}
>
  <span>
    id 1, description=location1
  </span>
  <br />
  <Form>
    <FormGroup
      fieldId="popup-form"
      label="Description"
    >
      <ForwardRef
        className="popup-input"
        id="input-description"
        onChange={[Function]}
        type="text"
        value="location1"
      />
    </FormGroup>
    <ActionGroup>
      <Component
        id="save-button"
        onClick={[Function]}
        variant="primary"
      >
        Save
      </Component>
      <Component
        id="remove-button"
        onClick={[Function]}
        variant="link"
      >
        Remove
      </Component>
      <Component
        id="cancel-button"
        onClick={[Function]}
        variant="link"
      >
        Cancel
      </Component>
    </ActionGroup>
  </Form>
</ForwardRef(Leaflet(Popup))>
`;
